### DISCUSSION TITLE
Lesson 7 : brownie.exceptions.VirtualMachineError: revert

### DISCUSSION THREAD

**Discussion by mbaurin** [smartcontractkit/full-blockchain-solidity-course-py#1029] [Q&A]
Hello I have an issue with an error that I do not understand : brownie.exceptions.VirtualMachineError: revert

 Here is my code and the error message : 

<img width="582" alt="Capture d’écran 2022-02-11 à 15 40 32" src="https://user-images.githubusercontent.com/86267428/153611987-99e2edcd-5862-4fad-926d-95967d178543.png">

`//SPDX-License-Indentifier: MIT

pragma solidity ^0.6.6;

import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";
import "@openzeppelin/contracts/access/Ownable.sol";

contract Lottery {

    address payable[] public players;
    uint256 public usdEntryFee;
    AggregatorV3Interface internal ethUsdPriceFeed;

    constructor(address _priceFeedAddress) public {
        usdEntryFee = 50 * (10**18);
        ethUsdPriceFeed = AggregatorV3Interface(_priceFeedAddress);
    } 

    function enter() public payable {
        // 50$ dollar min to participate
        players.push(msg.sender);
    }

    function getEntranceFee() public view returns (uint256) {
        (, int256 price, , , ) = ethUsdPriceFeed.latestRoundData();
        uint256 adjustedPrice = uint256(price) * 10**10; // 18 decimal
        uint256 costToEnter = (usdEntryFee * 10**18) / adjustedPrice;
        return costToEnter;
    }

    function startLottery() public {}

    function endLottery() public {}

}`

<img width="542" alt="Capture d’écran 2022-02-11 à 15 40 40" src="https://user-images.githubusercontent.com/86267428/153612051-2f4bb0d1-d830-4f00-b370-20be4d3da0cc.png">

`from brownie import Lottery, accounts, config, network
from web3 import Web3

def test_get_entrance_fee():
    account = accounts[0]
    lottery = Lottery.deploy(
        config["networks"][network.show_active()]["eth_usd_price_feed"],
        {"from": account},
    )
    assert lottery.getEntranceFee() > Web3.toWei(0.015, "ether")
    assert lottery.getEntranceFee() < Web3.toWei(0.019, "ether")`

<img width="551" alt="Capture d’écran 2022-02-11 à 15 40 46" src="https://user-images.githubusercontent.com/86267428/153612105-e4c86afc-7e9c-4380-83cf-da10280238f3.png">

`dependencies:
  - smartcontractkit/chainlink-brownie-contracts@1.1.1
  - OpenZeppelin/openzeppelin-contracts@3.4.0
compiler:
  solc:
    remappings:
      - '@chainlink=smartcontractkit/chainlink-brownie-contracts@1.1.1'
      - '@openzeppelin=OpenZeppelin/openzeppelin-contracts@3.4.0'
networks:
  mainnet-fork:
    eth_usd_price_feed: '0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419'
`

<img width="1203" alt="Capture d’écran 2022-02-11 à 15 41 27" src="https://user-images.githubusercontent.com/86267428/153612125-7acfd5d0-4a0b-44ed-89af-dbbdbb735ae1.png">

**Comment by n4n0b1t3** [comment]
Could you please paste the code as text and format the whole block with <> as source code.
I don't recognize the code of the test, is that from the video?

**Reply by mbaurin** [reply]
Just edited it, and yes it is from the video at the start of the lottery building

**Reply by n4n0b1t3** [reply]
OK, next time just use the editor tool <> this will insert triple backticks for multiline code snippets and won't mess up the code.
I will check the code now.

**Reply by n4n0b1t3** [reply]
1. config file is very different from the example in the repository, e.g. dotenv: .env is missing: https://github.com/PatrickAlphaC/smartcontract-lottery/blob/main/brownie-config.yaml
2. The test file is as well very different from the one in the repo: https://github.com/PatrickAlphaC/smartcontract-lottery/blob/main/tests/test_lottery_unit.py

I would suggest just trying the whole thing with the code in the repository and from there reduce and deduct what might be wrong in your case. But first, look into the .env issue. I am quite sure it is missing.

**Reply by cromewar** [reply]
@n4n0b1t3 The reasons why is different is because he is a o early time stamp of the video on where the test is like that.
@mbaurin  Could you please share this two things:

1. The command  you are using for testing
2. your brownie-config.yaml

**Reply by mbaurin** [reply]
Hi thanks for the help @cromewar ,

I am using the command : brownie test --network mainnet-fork

And my brownie-config file look like this : 

`
```YAML
dependencies:
  - smartcontractkit/chainlink-brownie-contracts@1.1.1
  - OpenZeppelin/openzeppelin-contracts@3.4.0

compiler:
  solc:
    remappings:
      - '@chainlink=smartcontractkit/chainlink-brownie-contracts@1.1.1'
      - '@OpenZeppelin=OpenZeppelin/openzeppelin-contracts@3.4.0'
      
networks:
  mainnet-fork:
    eth_usd_price_feed: '0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419'
```
`

**Reply by mbaurin** [reply]
Also after retrying the test today I am getting a different error that look like this :

<img width="1191" alt="Capture d’écran 2022-02-12 à 13 16 23" src="https://user-images.githubusercontent.com/86267428/153710927-69142920-3c08-4332-bda2-3e53e6c8d938.png">

**Reply by cromewar** [reply]
@mbaurin some tips here:

1. don't use python v 3.10, has problems with web3 libraries, is recommended to use python 3.8 instead.
2. Make sure you hace installed ganache as it seems to not recognize it.
3. be sure to select the correct python interpreter on vscode (bottom left corner).
